hudi-utilities/src/main/java/org/apache/hudi/utilities/schema/converter/JsonToAvroSchemaConverter.java

```diff
@@ -0,0 +1,250 @@
+        .put("name", sanitizeAsAvroName(name))
+        .put("doc", getAvroDoc(jsonProperty));
+
+    // infer `default`
+    boolean nullable = !isRequired;
+    if (jsonProperty.has("default")) {
+      avroNode.set("default", jsonProperty.get("default"));
+    } else if (nullable) {
+      avroNode.set("default", NullNode.getInstance());
+    }
+
+    // infer `types`
+    Set<String> avroTypeSet = new HashSet<>();
+    if (jsonProperty.hasNonNull("oneOf") || jsonProperty.hasNonNull("allOf")) {
+      // prefer to look for `oneOf` and `allOf` for types
+      Option<JsonNode> oneOfTypes = Option.ofNullable(jsonProperty.get("oneOf"));
+      if (oneOfTypes.isPresent()) {
+        oneOfTypes.get().elements().forEachRemaining(e -> avroTypeSet.add(JSON_TO_AVRO_TYPE.get(e.get("type").asText())));
+      }
+      Option<JsonNode> allOfTypes = Option.ofNullable(jsonProperty.get("allOf"));
+      if (allOfTypes.isPresent()) {
+        allOfTypes.get().elements().forEachRemaining(e -> avroTypeSet.add(JSON_TO_AVRO_TYPE.get(e.get("type").asText())));
+      }
+    } else if (jsonProperty.has("type")) {
+      // fall back to `type` parameter
+      JsonNode jsonType = jsonProperty.get("type");
+      if (jsonType.isArray()) {
+        jsonType.elements().forEachRemaining(e -> avroTypeSet.add(JSON_TO_AVRO_TYPE.get(e.asText())));
+      } else {
+        avroTypeSet.add(JSON_TO_AVRO_TYPE.get(jsonType.asText()));
+      }
+    }
+    List<String> avroTypes = new ArrayList<>();
+    if (nullable || avroTypeSet.contains("null")) {
+      avroTypes.add("null");
+    }
+    avroTypeSet.remove("null");
+    avroTypes.addAll(avroTypeSet);
+    avroNode.set("type", avroTypes.size() > 1
+        ? MAPPER.createArrayNode().addAll(avroTypes.stream().map(TextNode::valueOf).collect(Collectors.toList()))
+        : TextNode.valueOf(avroTypes.get(0)));
+    return avroNode;
+  }
+
+  private static boolean isJsonNestedType(JsonNode jsonNode) {
+    return jsonNode.has("type") && Objects.equals(jsonNode.get("type").asText(), "object");
+  }
+
+  private static boolean isJsonArrayType(JsonNode jsonNode) {
+    return jsonNode.has("type") && Objects.equals(jsonNode.get("type").asText(), "array");
+  }
+
+  private static boolean isJsonEnumType(JsonNode jsonNode) {
+    return jsonNode.hasNonNull("enum") && jsonNode.get("enum").isArray();
+  }
+
+  private static Option<String> getAvroSchemaRecordNamespace(JsonNode jsonNode) {
+    if (jsonNode.hasNonNull("$id")) {
+      return Option.of(URI.create(jsonNode.get("$id").asText()).getHost());
+    }
+    return Option.empty();
+  }
+
+  private static String getAvroSchemaRecordName(JsonNode jsonNode) {
+    if (jsonNode.hasNonNull("title")) {
+      return sanitizeAsAvroName(jsonNode.get("title").asText());
+    }
+    if (jsonNode.hasNonNull("$id")) {
+      // infer name from host: http://www.my-example.com => "my_example"
+      String host = URI.create(jsonNode.get("$id").asText()).getHost();
```

Comment:
Does `URI.create()` automatically convert hyphen to underscore?

Reply:
the comment here is meant for the if block not just the URI.create